diff --git a/README.md b/README.md
index 07860f0..f939bdc 100644
--- a/README.md
+++ b/README.md
@@ -1,8 +1,9 @@
+This is a fork of Nanopb version 0.3.9 from https://github.com/nanopb/nanopb updated with fixes to make it workable on Ledger Nano S.
+Thanks to https://github.com/yatendra for doing most of the work.
+
 Nanopb - Protocol Buffers for Embedded Systems
 ==============================================
 
-[![Build Status](https://travis-ci.org/nanopb/nanopb.svg?branch=master)](https://travis-ci.org/nanopb/nanopb)
-
 Nanopb is a small code-size Protocol Buffers implementation in ansi C. It is
 especially suitable for use in microcontrollers, but fits any memory
 restricted system.
diff --git a/generator/nanopb_generator.py b/generator/nanopb_generator.py
index 057df40..186c99f 100755
--- a/generator/nanopb_generator.py
+++ b/generator/nanopb_generator.py
@@ -1516,7 +1516,7 @@ def parse_file(filename, fdesc, options):
             optfilename = os.path.join(p, optfilename)
             if options.verbose:
                 sys.stderr.write('Reading options from ' + optfilename + '\n')
-            Globals.separate_options = read_options_file(open(optfilename, "rU"))
+            Globals.separate_options = read_options_file(open(optfilename, "r"))
             break
     else:
         # If we are given a full filename and it does not exist, give an error.
diff --git a/pb.h b/pb.h
index 64c0b43..fce1f1b 100644
--- a/pb.h
+++ b/pb.h
@@ -11,7 +11,7 @@
  *****************************************************************/
 
 /* Enable support for dynamically allocated fields */
-/* #define PB_ENABLE_MALLOC 1 */
+/*#define PB_ENABLE_MALLOC 1 */
 
 /* Define this if your CPU / compiler combination does not support
  * unaligned memory access to packed structures. */
@@ -25,7 +25,7 @@
 /* #define PB_FIELD_16BIT 1 */
 
 /* Add support for tag numbers > 65536 and fields larger than 65536 bytes. */
-/* #define PB_FIELD_32BIT 1 */
+#define PB_FIELD_32BIT 1
 
 /* Disable support for error messages in order to save some code space. */
 /* #define PB_NO_ERRMSG 1 */
@@ -46,7 +46,7 @@
 
 /* Version of the nanopb library. Just in case you want to check it in
  * your own program. */
-#define NANOPB_VERSION nanopb-0.3.9
+#define NANOPB_VERSION nanopb-0.4.0-dev
 
 /* Include all the system headers needed by nanopb. You will need the
  * definitions of the following:
@@ -571,7 +571,14 @@ struct pb_extension_s {
  * PB_GET_ERROR() always returns a pointer to a string.
  * PB_RETURN_ERROR() sets the error and returns false from current
  *                   function.
+ * G_depth is a global that should be initialized to 0 when 
+ * creating an input stream. It helps tracking the level of the 
+ * call stack when instrumentation is enabled 
+ * (set `INTRUMENT_FUNCTION += -finstrument-functions` in your 
+ * Makfile to do so)
  */
+extern int G_depth;
+
 #ifdef PB_NO_ERRMSG
 #define PB_SET_ERROR(stream, msg) PB_UNUSED(stream)
 #define PB_GET_ERROR(stream) "(errmsg disabled)"
diff --git a/pb_common.c b/pb_common.c
index 4fb7186..adfa747 100644
--- a/pb_common.c
+++ b/pb_common.c
@@ -4,22 +4,23 @@
  */
 
 #include "pb_common.h"
+#include "os.h"
+
 
 bool pb_field_iter_begin(pb_field_iter_t *iter, const pb_field_t *fields, void *dest_struct)
 {
-    iter->start = fields;
-    iter->pos = fields;
+    iter->start = (const pb_field_t *)PIC(fields);
+    iter->pos = (const pb_field_t *)PIC(fields);
     iter->required_field_index = 0;
     iter->dest_struct = dest_struct;
     iter->pData = (char*)dest_struct + iter->pos->data_offset;
     iter->pSize = (char*)iter->pData + iter->pos->size_offset;
-    
     return (iter->pos->tag != 0);
 }
 
 bool pb_field_iter_next(pb_field_iter_t *iter)
 {
-    const pb_field_t *prev_field = iter->pos;
+    const pb_field_t *prev_field = ((const pb_field_t *)PIC(iter->pos));
 
     if (prev_field->tag == 0)
     {
@@ -28,12 +29,12 @@ bool pb_field_iter_next(pb_field_iter_t *iter)
         return false;
     }
     
-    iter->pos++;
+    iter->pos = ((const pb_field_t *)PIC(iter->pos))+1;
     
-    if (iter->pos->tag == 0)
+    if (((const pb_field_t *)PIC(iter->pos))->tag == 0)
     {
         /* Wrapped back to beginning, reinitialize */
-        (void)pb_field_iter_begin(iter, iter->start, iter->dest_struct);
+        (void)pb_field_iter_begin(iter, ((const pb_field_t *)PIC(iter->start)), iter->dest_struct);
         return false;
     }
     else
@@ -42,8 +43,8 @@ bool pb_field_iter_next(pb_field_iter_t *iter)
         size_t prev_size = prev_field->data_size;
     
         if (PB_HTYPE(prev_field->type) == PB_HTYPE_ONEOF &&
-            PB_HTYPE(iter->pos->type) == PB_HTYPE_ONEOF &&
-            iter->pos->data_offset == PB_SIZE_MAX)
+            PB_HTYPE(((const pb_field_t *)PIC(iter->pos))->type) == PB_HTYPE_ONEOF &&
+            ((const pb_field_t *)PIC(iter->pos))->data_offset == PB_SIZE_MAX)
         {
             /* Don't advance pointers inside unions */
             return true;
@@ -69,29 +70,41 @@ bool pb_field_iter_next(pb_field_iter_t *iter)
             iter->required_field_index++;
         }
     
-        iter->pData = (char*)iter->pData + prev_size + iter->pos->data_offset;
-        iter->pSize = (char*)iter->pData + iter->pos->size_offset;
+        iter->pData = (char*)iter->pData + prev_size + ((const pb_field_t *)PIC(iter->pos))->data_offset;
+        iter->pSize = (char*)iter->pData + ((const pb_field_t *)PIC(iter->pos))->size_offset;
         return true;
     }
 }
 
 bool pb_field_iter_find(pb_field_iter_t *iter, uint32_t tag)
 {
-    const pb_field_t *start = iter->pos;
+    const pb_field_t *start = ((const pb_field_t *)PIC(iter->pos));
     
     do {
-        if (iter->pos->tag == tag &&
-            PB_LTYPE(iter->pos->type) != PB_LTYPE_EXTENSION)
+        if (((const pb_field_t *)PIC(iter->pos))->tag == tag &&
+            PB_LTYPE(((const pb_field_t *)PIC(iter->pos))->type) != PB_LTYPE_EXTENSION)
         {
             /* Found the wanted field */
             return true;
         }
         
         (void)pb_field_iter_next(iter);
-    } while (iter->pos != start);
+    } while (((const pb_field_t *)PIC(iter->pos)) != start);
     
     /* Searched all the way back to start, and found nothing. */
     return false;
 }
 
+extern void _ebss;
 
+//Return true if there is less than MIN_BSS_STACK_GAP bytes available in the stack
+bool check_stack_overflow(uint16_t min_bss_stack_gap){
+    uint32_t stack_top = 0;
+
+    //PRINTF("+++++++Stack remaining: CUR STACK ADDR: %p, EBSS: %p, diff: %d\n", &stack_top, &_ebss, ((uintptr_t)&stack_top) - ((uintptr_t)&_ebss));
+    if(((uintptr_t)&stack_top) < (((uintptr_t)&_ebss) + min_bss_stack_gap)){
+        return true;
+    }
+    
+    return false;
+}
diff --git a/pb_decode.c b/pb_decode.c
index dc344dc..17cc3e9 100644
--- a/pb_decode.c
+++ b/pb_decode.c
@@ -17,6 +17,8 @@
 #include "pb_decode.h"
 #include "pb_common.h"
 
+#include "os.h"
+
 /**************************************
  * Declarations internal to this file *
  **************************************/
@@ -27,7 +29,7 @@ static bool checkreturn buf_read(pb_istream_t *stream, pb_byte_t *buf, size_t co
 static bool checkreturn read_raw_value(pb_istream_t *stream, pb_wire_type_t wire_type, pb_byte_t *buf, size_t *size);
 static bool checkreturn decode_static_field(pb_istream_t *stream, pb_wire_type_t wire_type, pb_field_iter_t *iter);
 static bool checkreturn decode_callback_field(pb_istream_t *stream, pb_wire_type_t wire_type, pb_field_iter_t *iter);
-static bool checkreturn decode_field(pb_istream_t *stream, pb_wire_type_t wire_type, pb_field_iter_t *iter);
+static bool inline checkreturn decode_field(pb_istream_t *stream, pb_wire_type_t wire_type, pb_field_iter_t *iter) __attribute__((no_instrument_function));
 static void iter_from_extension(pb_field_iter_t *iter, pb_extension_t *extension);
 static bool checkreturn default_extension_decoder(pb_istream_t *stream, pb_extension_t *extension, uint32_t tag, pb_wire_type_t wire_type);
 static bool checkreturn decode_extension(pb_istream_t *stream, uint32_t tag, pb_wire_type_t wire_type, pb_field_iter_t *iter);
@@ -35,7 +37,7 @@ static bool checkreturn find_extension_field(pb_field_iter_t *iter);
 static void pb_field_set_to_default(pb_field_iter_t *iter);
 static void pb_message_set_to_defaults(const pb_field_t fields[], void *dest_struct);
 static bool checkreturn pb_dec_varint(pb_istream_t *stream, const pb_field_t *field, void *dest);
-static bool checkreturn pb_decode_varint32_eof(pb_istream_t *stream, uint32_t *dest, bool *eof);
+static bool inline checkreturn pb_decode_varint32_eof(pb_istream_t *stream, uint32_t *dest, bool *eof) __attribute__((no_instrument_function));
 static bool checkreturn pb_dec_uvarint(pb_istream_t *stream, const pb_field_t *field, void *dest);
 static bool checkreturn pb_dec_svarint(pb_istream_t *stream, const pb_field_t *field, void *dest);
 static bool checkreturn pb_dec_fixed32(pb_istream_t *stream, const pb_field_t *field, void *dest);
@@ -51,6 +53,7 @@ static bool checkreturn pb_skip_string(pb_istream_t *stream);
 static bool checkreturn allocate_field(pb_istream_t *stream, void *pData, size_t data_size, size_t array_size);
 static bool checkreturn pb_release_union_field(pb_istream_t *stream, pb_field_iter_t *iter);
 static void pb_release_single_field(const pb_field_iter_t *iter);
+static void initialize_pointer_field(void *pItem, pb_field_iter_t *iter);
 #endif
 
 #ifdef PB_WITHOUT_64BIT
@@ -78,6 +81,105 @@ static const pb_decoder_t PB_DECODERS[PB_LTYPES_COUNT] = {
     &pb_dec_fixed_length_bytes
 };
 
+extern void _ebss;
+extern void _estack;
+int G_depth;
+int last_stack_left;
+typedef struct addr_to_fname_s {
+    const void* const addr;
+    const char* const name;
+} addr_to_fname_t;
+
+static bool checkreturn inline decode_pointer_field(pb_istream_t *stream, pb_wire_type_t wire_type, pb_field_iter_t *iter) __attribute__((no_instrument_function));
+static bool checkreturn pb_readbyte(pb_istream_t *stream, pb_byte_t *buf);
+
+const addr_to_fname_t addr_to_fnames[] = {
+    { pb_readbyte, "pb_readbyte" },
+    { decode_pointer_field, "decode_pointer_field" },
+#ifdef PB_ENABLE_MALLOC
+    { initialize_pointer_field, "initialize_pointer_field" },
+    { allocate_field, "allocate_field" },
+    { pb_release_union_field, "pb_release_union_field" },
+    { pb_release_single_field, "pb_release_single_field" },
+    { pb_release, "pb_release" },
+#endif
+    { pb_decode, "pb_decode" },
+    { pb_message_set_to_defaults, "pb_message_set_to_defaults" },
+    { pb_decode_noinit, "pb_decode_noinit" },
+    { buf_read, "buf_read" },
+    { read_raw_value, "read_raw_value" },
+    { decode_static_field, "decode_static_field" },
+    { decode_callback_field, "decode_callback_field" },
+    { decode_field, "decode_field" },
+    { iter_from_extension, "iter_from_extension" },
+    { default_extension_decoder, "default_extension_decoder" },
+    { decode_extension, "decode_extension" },
+    { find_extension_field, "find_extension_field" },
+    { pb_field_set_to_default, "pb_field_set_to_default" },
+    { pb_message_set_to_defaults, "pb_message_set_to_defaults" },
+    { pb_dec_varint, "pb_dec_varint" },
+    { pb_decode_varint32_eof, "pb_decode_varint32_eof" },
+    { pb_dec_uvarint, "pb_dec_uvarint" },
+    { pb_dec_svarint, "pb_dec_svarint" },
+    { pb_dec_fixed32, "pb_dec_fixed32" },
+    { pb_dec_fixed64, "pb_dec_fixed64" },
+    { pb_dec_bytes, "pb_dec_bytes" },
+    { pb_dec_string, "pb_dec_string" },
+    { pb_dec_submessage, "pb_dec_submessage" },
+    { pb_dec_fixed_length_bytes, "pb_dec_fixed_length_bytes" },
+    { pb_skip_varint, "pb_skip_varint" },
+    { pb_skip_string, "pb_skip_string" },
+    { pb_decode_delimited, "pb_decode_delimited" },
+    { pb_decode_delimited_noinit, "pb_decode_delimited_noinit" },
+    { pb_decode_nullterminated, "pb_decode_nullterminated" },
+    { pb_istream_from_buffer, "pb_istream_from_buffer" },
+    { pb_read, "pb_read" },
+    { pb_decode_tag, "pb_decode_tag" },
+    { pb_skip_field, "pb_skip_field" },
+    // { pb_decode_varint, "pb_decode_varint" },
+    { pb_decode_varint32, "pb_decode_varint32" },
+    { pb_decode_svarint, "pb_decode_svarint" },
+    { pb_decode_fixed32, "pb_decode_fixed32" },
+    { pb_make_string_substream, "pb_make_string_substream" },
+    { pb_close_string_substream, "pb_close_string_substream" },
+    { decode_pointer_field, "decode_pointer_field" },
+    { decode_callback_field, "decode_callback_field" }
+};
+
+const char* addr_to_fname(void* func) __attribute__((no_instrument_function));
+const char* addr_to_fname(void* func){
+    for(int i = 0; i < sizeof(addr_to_fnames)/sizeof(*addr_to_fnames); i++){
+        if(PIC(addr_to_fnames[i].addr) == func){
+            return (const char*)(PIC(addr_to_fnames[i].name));
+        }
+    }
+    return "";
+}
+
+void __cyg_profile_func_enter(void *this_fn, void *call_site) __attribute__((no_instrument_function));
+void __cyg_profile_func_enter( void *func, void *callsite )
+{
+    const char* fname = addr_to_fname(func);
+    for(int i = 0; i < G_depth; i++){
+        PRINTF(" ");
+    }
+    PRINTF("-> %p '%s' [from %p], STACK %p, left: %d\n", func, fname, callsite, &fname, ((void*)&fname) - &_ebss); //, last_stack_left - (((void*)&fname) - &_ebss));
+    last_stack_left = ((void*)&fname) - &_ebss;
+    G_depth++;
+}
+void __cyg_profile_func_exit(void *this_fn, void *call_site) __attribute__((no_instrument_function));
+void __cyg_profile_func_exit( void *func, void *callsite )
+{
+    const char* fname = addr_to_fname(func);
+    // last_stack_left = ((void*)&fname) - &_ebss;
+    G_depth--;
+    for(int i = 0; i < G_depth; i++){
+        PRINTF(" ");
+    }
+    PRINTF("<- %p '%s' [from %p], left: %d\n", func, fname, callsite, ((void*)&fname) - &_ebss);
+    // last_stack_left = ((void*)&fname) - &_ebss;
+}
+
 /*******************************
  * pb_istream_t implementation *
  *******************************/
@@ -260,7 +362,6 @@ bool checkreturn pb_decode_varint(pb_istream_t *stream, uint64_t *dest)
         
         if (!pb_readbyte(stream, &byte))
             return false;
-
         result |= (uint64_t)(byte & 0x7F) << bitpos;
         bitpos = (uint_fast8_t)(bitpos + 7);
     } while (byte & 0x80);
@@ -310,6 +411,7 @@ bool checkreturn pb_decode_tag(pb_istream_t *stream, pb_wire_type_t *wire_type,
     
     *tag = temp >> 3;
     *wire_type = (pb_wire_type_t)(temp & 7);
+    
     return true;
 }
 
@@ -392,28 +494,30 @@ bool checkreturn pb_close_string_substream(pb_istream_t *stream, pb_istream_t *s
  * Decode a single field *
  *************************/
 
+pb_type_t type;
+pb_decoder_t func;
 static bool checkreturn decode_static_field(pb_istream_t *stream, pb_wire_type_t wire_type, pb_field_iter_t *iter)
 {
-    pb_type_t type;
-    pb_decoder_t func;
-    
-    type = iter->pos->type;
-    func = PB_DECODERS[PB_LTYPE(type)];
-
+    type = ((const pb_field_t *)PIC(iter->pos))->type;
+    func = (pb_decoder_t)PIC(PB_DECODERS[PB_LTYPE(type)]);
+                  
     switch (PB_HTYPE(type))
     {
         case PB_HTYPE_REQUIRED:
-            return func(stream, iter->pos, iter->pData);
+            return func(stream, ((const pb_field_t *)PIC(iter->pos)), iter->pData);
             
         case PB_HTYPE_OPTIONAL:
             if (iter->pSize != iter->pData)
                 *(bool*)iter->pSize = true;
-            return func(stream, iter->pos, iter->pData);
+            return func(stream, ((const pb_field_t *)PIC(iter->pos)), iter->pData);
     
         case PB_HTYPE_REPEATED:
             if (wire_type == PB_WT_STRING
                 && PB_LTYPE(type) <= PB_LTYPE_LAST_PACKABLE)
             {
+#ifndef PROTO2_SUPPORT
+                return false;
+#else
                 /* Packed array */
                 bool status = true;
                 pb_size_t *size = (pb_size_t*)iter->pSize;
@@ -421,10 +525,10 @@ static bool checkreturn decode_static_field(pb_istream_t *stream, pb_wire_type_t
                 if (!pb_make_string_substream(stream, &substream))
                     return false;
                 
-                while (substream.bytes_left > 0 && *size < iter->pos->array_size)
+                while (substream.bytes_left > 0 && *size < ((const pb_field_t *)PIC(iter->pos))->array_size)
                 {
-                    void *pItem = (char*)iter->pData + iter->pos->data_size * (*size);
-                    if (!func(&substream, iter->pos, pItem))
+                    void *pItem = (char*)iter->pData + ((const pb_field_t *)PIC(iter->pos))->data_size * (*size);
+                    if (!func(&substream, ((const pb_field_t *)PIC(iter->pos)), pItem))
                     {
                         status = false;
                         break;
@@ -438,29 +542,30 @@ static bool checkreturn decode_static_field(pb_istream_t *stream, pb_wire_type_t
                     return false;
 
                 return status;
+#endif
             }
             else
             {
                 /* Repeated field */
                 pb_size_t *size = (pb_size_t*)iter->pSize;
-                void *pItem = (char*)iter->pData + iter->pos->data_size * (*size);
-                if (*size >= iter->pos->array_size)
+                void *pItem = (char*)iter->pData + ((const pb_field_t *)PIC(iter->pos))->data_size * (*size);
+                if (*size >= ((const pb_field_t *)PIC(iter->pos))->array_size)
                     PB_RETURN_ERROR(stream, "array overflow");
                 
                 (*size)++;
-                return func(stream, iter->pos, pItem);
+                return func(stream, ((const pb_field_t *)PIC(iter->pos)), pItem);
             }
 
         case PB_HTYPE_ONEOF:
-            *(pb_size_t*)iter->pSize = iter->pos->tag;
+            *(pb_size_t*)iter->pSize = ((const pb_field_t *)PIC(iter->pos))->tag;
             if (PB_LTYPE(type) == PB_LTYPE_SUBMESSAGE)
             {
                 /* We memset to zero so that any callbacks are set to NULL.
                  * Then set any default values. */
-                memset(iter->pData, 0, iter->pos->data_size);
-                pb_message_set_to_defaults((const pb_field_t*)iter->pos->ptr, iter->pData);
+                memset(iter->pData, 0, ((const pb_field_t *)PIC(iter->pos))->data_size);
+                pb_message_set_to_defaults((const pb_field_t*)((const pb_field_t *)PIC(iter->pos))->ptr, iter->pData);
             }
-            return func(stream, iter->pos, iter->pData);
+            return func(stream, ((const pb_field_t *)PIC(iter->pos)), iter->pData);
 
         default:
             PB_RETURN_ERROR(stream, "invalid field type");
@@ -502,7 +607,6 @@ static bool checkreturn allocate_field(pb_istream_t *stream, void *pData, size_t
     ptr = pb_realloc(ptr, array_size * data_size);
     if (ptr == NULL)
         PB_RETURN_ERROR(stream, "realloc failed");
-    
     *(void**)pData = ptr;
     return true;
 }
@@ -510,17 +614,18 @@ static bool checkreturn allocate_field(pb_istream_t *stream, void *pData, size_t
 /* Clear a newly allocated item in case it contains a pointer, or is a submessage. */
 static void initialize_pointer_field(void *pItem, pb_field_iter_t *iter)
 {
-    if (PB_LTYPE(iter->pos->type) == PB_LTYPE_STRING ||
-        PB_LTYPE(iter->pos->type) == PB_LTYPE_BYTES)
+    if (PB_LTYPE(((const pb_field_t *)PIC(iter->pos))->type) == PB_LTYPE_STRING ||
+        PB_LTYPE(((const pb_field_t *)PIC(iter->pos))->type) == PB_LTYPE_BYTES)
     {
         *(void**)pItem = NULL;
     }
-    else if (PB_LTYPE(iter->pos->type) == PB_LTYPE_SUBMESSAGE)
+    else if (PB_LTYPE(((const pb_field_t *)PIC(iter->pos))->type) == PB_LTYPE_SUBMESSAGE)
     {
         /* We memset to zero so that any callbacks are set to NULL.
          * Then set any default values. */
-        memset(pItem, 0, iter->pos->data_size);
-        pb_message_set_to_defaults((const pb_field_t *) iter->pos->ptr, pItem);
+        memset(pItem, 0, ((const pb_field_t *)PIC(iter->pos))->data_size);
+
+        pb_message_set_to_defaults((const pb_field_t *) ((const pb_field_t *)PIC(iter->pos))->ptr, pItem);
     }
 }
 #endif
@@ -535,8 +640,8 @@ static bool checkreturn decode_pointer_field(pb_istream_t *stream, pb_wire_type_
     pb_type_t type;
     pb_decoder_t func;
     
-    type = iter->pos->type;
-    func = PB_DECODERS[PB_LTYPE(type)];
+    type = ((const pb_field_t *)PIC(iter->pos))->type;
+    func = PIC(PB_DECODERS[PB_LTYPE(type)]);
     
     switch (PB_HTYPE(type))
     {
@@ -546,27 +651,31 @@ static bool checkreturn decode_pointer_field(pb_istream_t *stream, pb_wire_type_
             if (PB_LTYPE(type) == PB_LTYPE_SUBMESSAGE &&
                 *(void**)iter->pData != NULL)
             {
+                // if an error has been observed, no need to release memory
+                if(stream->errmsg){
+                    return false;
+                }
                 /* Duplicate field, have to release the old allocation first. */
                 pb_release_single_field(iter);
             }
         
             if (PB_HTYPE(type) == PB_HTYPE_ONEOF)
             {
-                *(pb_size_t*)iter->pSize = iter->pos->tag;
+                *(pb_size_t*)iter->pSize = ((const pb_field_t *)PIC(iter->pos))->tag;
             }
 
             if (PB_LTYPE(type) == PB_LTYPE_STRING ||
                 PB_LTYPE(type) == PB_LTYPE_BYTES)
             {
-                return func(stream, iter->pos, iter->pData);
+                return func(stream, ((const pb_field_t *)PIC(iter->pos)), iter->pData);
             }
             else
             {
-                if (!allocate_field(stream, iter->pData, iter->pos->data_size, 1))
+                if (!allocate_field(stream, iter->pData, ((const pb_field_t *)PIC(iter->pos))->data_size, 1))
                     return false;
                 
                 initialize_pointer_field(*(void**)iter->pData, iter);
-                return func(stream, iter->pos, *(void**)iter->pData);
+                return func(stream, ((const pb_field_t *)PIC(iter->pos)), *(void**)iter->pData);
             }
     
         case PB_HTYPE_REPEATED:
@@ -590,9 +699,8 @@ static bool checkreturn decode_pointer_field(pb_istream_t *stream, pb_wire_type_
                         /* Allocate more storage. This tries to guess the
                          * number of remaining entries. Round the division
                          * upwards. */
-                        allocated_size += (substream.bytes_left - 1) / iter->pos->data_size + 1;
-                        
-                        if (!allocate_field(&substream, iter->pData, iter->pos->data_size, allocated_size))
+                        allocated_size += (substream.bytes_left - 1) / ((const pb_field_t *)PIC(iter->pos))->data_size + 1;
+                        if (!allocate_field(&substream, iter->pData, ((const pb_field_t *)PIC(iter->pos))->data_size, allocated_size))
                         {
                             status = false;
                             break;
@@ -600,9 +708,9 @@ static bool checkreturn decode_pointer_field(pb_istream_t *stream, pb_wire_type_
                     }
 
                     /* Decode the array entry */
-                    pItem = *(char**)iter->pData + iter->pos->data_size * (*size);
+                    pItem = *(char**)iter->pData + ((const pb_field_t *)PIC(iter->pos))->data_size * (*size);
                     initialize_pointer_field(pItem, iter);
-                    if (!func(&substream, iter->pos, pItem))
+                    if (!func(&substream, ((const pb_field_t *)PIC(iter->pos)), pItem))
                     {
                         status = false;
                         break;
@@ -634,12 +742,13 @@ static bool checkreturn decode_pointer_field(pb_istream_t *stream, pb_wire_type_
                     PB_RETURN_ERROR(stream, "too many array entries");
                 
                 (*size)++;
-                if (!allocate_field(stream, iter->pData, iter->pos->data_size, *size))
+                if (!allocate_field(stream, iter->pData, ((const pb_field_t *)PIC(iter->pos))->data_size, *size))
                     return false;
+
             
-                pItem = *(char**)iter->pData + iter->pos->data_size * (*size - 1);
+                pItem = *(char**)iter->pData + ((const pb_field_t *)PIC(iter->pos))->data_size * (*size - 1);
                 initialize_pointer_field(pItem, iter);
-                return func(stream, iter->pos, pItem);
+                return func(stream, ((const pb_field_t *)PIC(iter->pos)), pItem);
             }
 
         default:
@@ -656,6 +765,7 @@ static bool checkreturn decode_callback_field(pb_istream_t *stream, pb_wire_type
     void *arg = pCallback->arg;
 #else
     void **arg = &(pCallback->arg);
+
 #endif
     
     if (pCallback == NULL || pCallback->funcs.decode == NULL)
@@ -670,7 +780,7 @@ static bool checkreturn decode_callback_field(pb_istream_t *stream, pb_wire_type
         
         do
         {
-            if (!pCallback->funcs.decode(&substream, iter->pos, arg))
+            if (!pCallback->funcs.decode(&substream, ((const pb_field_t *)PIC(iter->pos)), arg))
                 PB_RETURN_ERROR(stream, "callback failed");
         } while (substream.bytes_left);
         
@@ -693,7 +803,7 @@ static bool checkreturn decode_callback_field(pb_istream_t *stream, pb_wire_type
             return false;
         substream = pb_istream_from_buffer(buffer, size);
         
-        return pCallback->funcs.decode(&substream, iter->pos, arg);
+        return pCallback->funcs.decode(&substream, ((const pb_field_t *)PIC(iter->pos)), arg);
     }
 }
 
@@ -702,14 +812,14 @@ static bool checkreturn decode_field(pb_istream_t *stream, pb_wire_type_t wire_t
 #ifdef PB_ENABLE_MALLOC
     /* When decoding an oneof field, check if there is old data that must be
      * released first. */
-    if (PB_HTYPE(iter->pos->type) == PB_HTYPE_ONEOF)
+    if (PB_HTYPE(((const pb_field_t *)PIC(iter->pos))->type) == PB_HTYPE_ONEOF)
     {
         if (!pb_release_union_field(stream, iter))
             return false;
     }
 #endif
-
-    switch (PB_ATYPE(iter->pos->type))
+    
+    switch (PB_ATYPE(((const pb_field_t *)PIC(iter->pos))->type))
     {
         case PB_ATYPE_STATIC:
             return decode_static_field(stream, wire_type, iter);
@@ -790,13 +900,13 @@ static bool checkreturn decode_extension(pb_istream_t *stream,
  * message. Returns false if no extension field is found. */
 static bool checkreturn find_extension_field(pb_field_iter_t *iter)
 {
-    const pb_field_t *start = iter->pos;
+    const pb_field_t *start = (const pb_field_t *)PIC(iter->pos);
     
     do {
-        if (PB_LTYPE(iter->pos->type) == PB_LTYPE_EXTENSION)
+        if (PB_LTYPE(((const pb_field_t *)PIC(iter->pos))->type) == PB_LTYPE_EXTENSION)
             return true;
         (void)pb_field_iter_next(iter);
-    } while (iter->pos != start);
+    } while (((const pb_field_t *)PIC(iter->pos)) != start);
     
     return false;
 }
@@ -805,8 +915,7 @@ static bool checkreturn find_extension_field(pb_field_iter_t *iter)
 static void pb_field_set_to_default(pb_field_iter_t *iter)
 {
     pb_type_t type;
-    type = iter->pos->type;
-    
+    type = ((const pb_field_t *)PIC(iter->pos))->type;
     if (PB_LTYPE(type) == PB_LTYPE_EXTENSION)
     {
         pb_extension_t *ext = *(pb_extension_t* const *)iter->pData;
@@ -839,28 +948,27 @@ static void pb_field_set_to_default(pb_field_iter_t *iter)
 
         if (init_data)
         {
-            if (PB_LTYPE(iter->pos->type) == PB_LTYPE_SUBMESSAGE)
+            if (PB_LTYPE(((const pb_field_t *)PIC(iter->pos))->type) == PB_LTYPE_SUBMESSAGE)
             {
                 /* Initialize submessage to defaults */
-                pb_message_set_to_defaults((const pb_field_t *) iter->pos->ptr, iter->pData);
+                pb_message_set_to_defaults((const pb_field_t *) ((const pb_field_t *)PIC(iter->pos))->ptr, iter->pData);
             }
-            else if (iter->pos->ptr != NULL)
+            else if (((const pb_field_t *)PIC(iter->pos))->ptr != NULL)
             {
                 /* Initialize to default value */
-                memcpy(iter->pData, iter->pos->ptr, iter->pos->data_size);
+                memcpy(iter->pData, ((const pb_field_t *)PIC(iter->pos))->ptr, ((const pb_field_t *)PIC(iter->pos))->data_size);
             }
             else
             {
                 /* Initialize to zeros */
-                memset(iter->pData, 0, iter->pos->data_size);
+                memset(iter->pData, 0, ((const pb_field_t *)PIC(iter->pos))->data_size);
             }
         }
     }
     else if (PB_ATYPE(type) == PB_ATYPE_POINTER)
-    {
+    {       
         /* Initialize the pointer to NULL. */
         *(void**)iter->pData = NULL;
-        
         /* Initialize array count to 0. */
         if (PB_HTYPE(type) == PB_HTYPE_REPEATED ||
             PB_HTYPE(type) == PB_HTYPE_ONEOF)
@@ -877,10 +985,9 @@ static void pb_field_set_to_default(pb_field_iter_t *iter)
 static void pb_message_set_to_defaults(const pb_field_t fields[], void *dest_struct)
 {
     pb_field_iter_t iter;
-
+                
     if (!pb_field_iter_begin(&iter, fields, dest_struct))
         return; /* Empty message type */
-    
     do
     {
         pb_field_set_to_default(&iter);
@@ -890,24 +997,25 @@ static void pb_message_set_to_defaults(const pb_field_t fields[], void *dest_str
 /*********************
  * Decode all fields *
  *********************/
-
+#define PB_DECODE_STACK_REQUIREMENT 260
 bool checkreturn pb_decode_noinit(pb_istream_t *stream, const pb_field_t fields[], void *dest_struct)
 {
-    uint32_t fields_seen[(PB_MAX_REQUIRED_FIELDS + 31) / 32] = {0, 0};
-    const uint32_t allbits = ~(uint32_t)0;
-    uint32_t extension_range_start = 0;
+    if(check_stack_overflow(PB_DECODE_STACK_REQUIREMENT))
+        PB_RETURN_ERROR(stream, "not enough stack");
+
+    // uint32_t fields_seen[(PB_MAX_REQUIRED_FIELDS + 31) / 32] = {0, 0};
+    // const uint32_t allbits = ~(uint32_t)0;
+    // uint32_t extension_range_start = 0;
     pb_field_iter_t iter;
     
     /* Return value ignored, as empty message types will be correctly handled by
      * pb_field_iter_find() anyway. */
     (void)pb_field_iter_begin(&iter, fields, dest_struct);
-    
     while (stream->bytes_left)
     {
         uint32_t tag;
         pb_wire_type_t wire_type;
         bool eof;
-        
         if (!pb_decode_tag(stream, &wire_type, &tag, &eof))
         {
             if (eof)
@@ -915,16 +1023,17 @@ bool checkreturn pb_decode_noinit(pb_istream_t *stream, const pb_field_t fields[
             else
                 return false;
         }
-        
         if (!pb_field_iter_find(&iter, tag))
         {
+
+#ifdef PROTO2_SUPPORT
             /* No match found, check if it matches an extension. */
             if (tag >= extension_range_start)
             {
                 if (!find_extension_field(&iter))
                     extension_range_start = (uint32_t)-1;
                 else
-                    extension_range_start = iter.pos->tag;
+                    extension_range_start = ((const pb_field_t*)PIC(iter.pos))->tag;
                 
                 if (tag >= extension_range_start)
                 {
@@ -940,24 +1049,26 @@ bool checkreturn pb_decode_noinit(pb_istream_t *stream, const pb_field_t fields[
                     }
                 }
             }
-        
+#endif
             /* No match found, skip data */
             if (!pb_skip_field(stream, wire_type))
                 return false;
             continue;
         }
-        
-        if (PB_HTYPE(iter.pos->type) == PB_HTYPE_REQUIRED
+
+#ifdef PROTO2_SUPPORT
+        if (PB_HTYPE(((const pb_field_t*)PIC(iter.pos))->type) == PB_HTYPE_REQUIRED
             && iter.required_field_index < PB_MAX_REQUIRED_FIELDS)
         {
             uint32_t tmp = ((uint32_t)1 << (iter.required_field_index & 31));
             fields_seen[iter.required_field_index >> 5] |= tmp;
         }
-            
+#endif    
         if (!decode_field(stream, wire_type, &iter))
             return false;
     }
-    
+
+#ifdef PROTO2_SUPPORT
     /* Check that all required fields were present. */
     {
         /* First figure out the number of required fields by
@@ -969,11 +1080,11 @@ bool checkreturn pb_decode_noinit(pb_istream_t *stream, const pb_field_t fields[
         unsigned i;
         do {
             req_field_count = iter.required_field_index;
-            last_type = iter.pos->type;
+            last_type = ((const pb_field_t*)PIC(iter.pos))->type;
         } while (pb_field_iter_next(&iter));
         
         /* Fixup if last field was also required. */
-        if (PB_HTYPE(last_type) == PB_HTYPE_REQUIRED && iter.pos->tag != 0)
+        if (PB_HTYPE(last_type) == PB_HTYPE_REQUIRED && ((const pb_field_t*)PIC(iter.pos))->tag != 0)
             req_field_count++;
         
         if (req_field_count > PB_MAX_REQUIRED_FIELDS)
@@ -999,7 +1110,8 @@ bool checkreturn pb_decode_noinit(pb_istream_t *stream, const pb_field_t fields[
             }
         }
     }
-    
+#endif
+
     return true;
 }
 
@@ -1008,12 +1120,12 @@ bool checkreturn pb_decode(pb_istream_t *stream, const pb_field_t fields[], void
     bool status;
     pb_message_set_to_defaults(fields, dest_struct);
     status = pb_decode_noinit(stream, fields, dest_struct);
-    
+/*
 #ifdef PB_ENABLE_MALLOC
     if (!status)
         pb_release(fields, dest_struct);
 #endif
-    
+*/
     return status;
 }
 
@@ -1024,7 +1136,6 @@ bool pb_decode_delimited_noinit(pb_istream_t *stream, const pb_field_t fields[],
 
     if (!pb_make_string_substream(stream, &substream))
         return false;
-
     status = pb_decode_noinit(&substream, fields, dest_struct);
 
     if (!pb_close_string_substream(stream, &substream))
@@ -1058,8 +1169,12 @@ bool pb_decode_nullterminated(pb_istream_t *stream, const pb_field_t fields[], v
  * release it before overwriting with a different one. */
 static bool pb_release_union_field(pb_istream_t *stream, pb_field_iter_t *iter)
 {
+    // if an error has been observed, no need to release memory
+    if(stream->errmsg){
+        return false;
+    }
     pb_size_t old_tag = *(pb_size_t*)iter->pSize; /* Previous which_ value */
-    pb_size_t new_tag = iter->pos->tag; /* New which_ value */
+    pb_size_t new_tag = ((const pb_field_t *)PIC(iter->pos))->tag; /* New which_ value */
 
     if (old_tag == 0)
         return true; /* Ok, no old data in union */
@@ -1085,11 +1200,11 @@ static bool pb_release_union_field(pb_istream_t *stream, pb_field_iter_t *iter)
 static void pb_release_single_field(const pb_field_iter_t *iter)
 {
     pb_type_t type;
-    type = iter->pos->type;
+    type = ((const pb_field_t *)PIC(iter->pos))->type;
 
     if (PB_HTYPE(type) == PB_HTYPE_ONEOF)
     {
-        if (*(pb_size_t*)iter->pSize != iter->pos->tag)
+        if (*(pb_size_t*)iter->pSize != ((const pb_field_t *)PIC(iter->pos))->tag)
             return; /* This is not the current field in the union */
     }
 
@@ -1123,10 +1238,10 @@ static void pb_release_single_field(const pb_field_iter_t *iter)
         {
             count = *(pb_size_t*)iter->pSize;
 
-            if (PB_ATYPE(type) == PB_ATYPE_STATIC && count > iter->pos->array_size)
+            if (PB_ATYPE(type) == PB_ATYPE_STATIC && count > ((const pb_field_t *)PIC(iter->pos))->array_size)
             {
                 /* Protect against corrupted _count fields */
-                count = iter->pos->array_size;
+                count = ((const pb_field_t *)PIC(iter->pos))->array_size;
             }
         }
         
@@ -1134,8 +1249,8 @@ static void pb_release_single_field(const pb_field_iter_t *iter)
         {
             while (count--)
             {
-                pb_release((const pb_field_t*)iter->pos->ptr, pItem);
-                pItem = (char*)pItem + iter->pos->data_size;
+                pb_release((const pb_field_t*)((const pb_field_t *)PIC(iter->pos))->ptr, pItem);
+                pItem = (char*)pItem + ((const pb_field_t *)PIC(iter->pos))->data_size;
             }
         }
     }
@@ -1235,14 +1350,13 @@ bool pb_decode_fixed64(pb_istream_t *stream, void *dest)
     return true;
 }
 #endif
-
 static bool checkreturn pb_dec_varint(pb_istream_t *stream, const pb_field_t *field, void *dest)
 {
-    pb_uint64_t value;
-    pb_int64_t svalue;
-    pb_int64_t clamped;
+    pb_uint64_t value=0;
+    pb_int64_t svalue=0;
+    pb_int64_t clamped=0;
     if (!pb_decode_varint(stream, &value))
-        return false;
+         return false;
     
     /* See issue 97: Google's C++ protobuf allows negative varint values to
      * be cast as int32_t, instead of the int64_t that should be used when
@@ -1275,7 +1389,7 @@ static bool checkreturn pb_dec_varint(pb_istream_t *stream, const pb_field_t *fi
 
 static bool checkreturn pb_dec_uvarint(pb_istream_t *stream, const pb_field_t *field, void *dest)
 {
-    pb_uint64_t value, clamped;
+    pb_uint64_t value = 0, clamped = 0;
     if (!pb_decode_varint(stream, &value))
         return false;
     
@@ -1414,7 +1528,7 @@ static bool checkreturn pb_dec_submessage(pb_istream_t *stream, const pb_field_t
 {
     bool status;
     pb_istream_t substream;
-    const pb_field_t* submsg_fields = (const pb_field_t*)field->ptr;
+    const pb_field_t* submsg_fields = (const pb_field_t*)PIC(field->ptr);
     
     if (!pb_make_string_substream(stream, &substream))
         return false;
@@ -1425,10 +1539,13 @@ static bool checkreturn pb_dec_submessage(pb_istream_t *stream, const pb_field_t
     /* New array entries need to be initialized, while required and optional
      * submessages have already been initialized in the top-level pb_decode. */
     if (PB_HTYPE(field->type) == PB_HTYPE_REPEATED)
+    {
         status = pb_decode(&substream, submsg_fields, dest);
+    }
     else
+    {
         status = pb_decode_noinit(&substream, submsg_fields, dest);
-    
+    }
     if (!pb_close_string_substream(stream, &substream))
         return false;
     return status;
diff --git a/pb_decode.h b/pb_decode.h
index 398b24a..5fcc201 100644
--- a/pb_decode.h
+++ b/pb_decode.h
@@ -12,6 +12,10 @@
 extern "C" {
 #endif
 
+/* Tracks call stack when instrumentation is enabled
+ * Should be set to 0 when instantiating pb_istream_t*/
+extern int G_depth;
+
 /* Structure for defining custom input streams. You will need to provide
  * a callback function to read the bytes from your storage, which can be
  * for example a file or a network socket.
@@ -45,6 +49,7 @@ struct pb_istream_s
 #endif
 };
 
+
 /***************************
  * Main decoding functions *
  ***************************/
@@ -105,7 +110,6 @@ bool pb_decode_nullterminated(pb_istream_t *stream, const pb_field_t fields[], v
 void pb_release(const pb_field_t fields[], void *dest_struct);
 #endif
 
-
 /**************************************
  * Functions for manipulating streams *
  **************************************/
@@ -122,7 +126,6 @@ pb_istream_t pb_istream_from_buffer(const pb_byte_t *buf, size_t bufsize);
  */
 bool pb_read(pb_istream_t *stream, pb_byte_t *buf, size_t count);
 
-
 /************************************************
  * Helper functions for writing field callbacks *
  ************************************************/
@@ -141,6 +144,7 @@ bool pb_decode_varint(pb_istream_t *stream, uint64_t *dest);
 #else
 #define pb_decode_varint pb_decode_varint32
 #endif
+// uint64_t val2;
 
 /* Decode an integer in the varint format. This works for bool, enum, int32,
  * and uint32 field types. */
diff --git a/pb_encode.c b/pb_encode.c
index e957d42..41d3c8a 100644
--- a/pb_encode.c
+++ b/pb_encode.c
@@ -7,6 +7,12 @@
 #include "pb_encode.h"
 #include "pb_common.h"
 
+#ifdef OS_IO_SEPROXYHAL
+#include "os.h"
+#else
+#define PIC(x) x
+#endif
+
 /* Use the GCC warn_unused_result attribute to check that all return values
  * are propagated correctly. On other compilers and gcc before 3.4.0 just
  * ignore the annotation.
@@ -309,7 +315,7 @@ static bool checkreturn encode_basic_field(pb_ostream_t *stream,
     bool implicit_has;
     const void *pSize = &implicit_has;
     
-    func = PB_ENCODERS[PB_LTYPE(field->type)];
+    func = PB_ENCODERS[PB_LTYPE(field->type)];    
     
     if (field->size_offset)
     {
@@ -344,7 +350,7 @@ static bool checkreturn encode_basic_field(pb_ostream_t *stream,
                 PB_RETURN_ERROR(stream, "missing required field");
             if (!pb_encode_tag_for_field(stream, field))
                 return false;
-            if (!func(stream, field, pData))
+            if (!((pb_encoder_t)PIC(func))(stream, field, pData))
                 return false;
             break;
         
@@ -353,8 +359,7 @@ static bool checkreturn encode_basic_field(pb_ostream_t *stream,
             {
                 if (!pb_encode_tag_for_field(stream, field))
                     return false;
-            
-                if (!func(stream, field, pData))
+                if (!((pb_encoder_t)PIC(func))(stream, field, pData))
                     return false;
             }
             break;
@@ -370,7 +375,7 @@ static bool checkreturn encode_basic_field(pb_ostream_t *stream,
                 if (!pb_encode_tag_for_field(stream, field))
                     return false;
 
-                if (!func(stream, field, pData))
+                if (!((pb_encoder_t)PIC(func))(stream, field, pData))
                     return false;
             }
             break;
@@ -378,7 +383,6 @@ static bool checkreturn encode_basic_field(pb_ostream_t *stream,
         default:
             PB_RETURN_ERROR(stream, "invalid field type");
     }
-    
     return true;
 }
 
@@ -486,8 +490,8 @@ bool checkreturn pb_encode(pb_ostream_t *stream, const pb_field_t fields[], cons
 {
     pb_field_iter_t iter;
     if (!pb_field_iter_begin(&iter, fields, pb_const_cast(src_struct)))
-        return true; /* Empty message type */
-    
+         return true; /* Empty message type */
+       
     do {
         if (PB_LTYPE(iter.pos->type) == PB_LTYPE_EXTENSION)
         {
@@ -545,6 +549,7 @@ bool checkreturn pb_encode_varint(pb_ostream_t *stream, pb_uint64_t value)
         pb_byte_t v = (pb_byte_t)value;
         return pb_write(stream, &v, 1);
     }
+
     
     while (value)
     {
@@ -553,7 +558,6 @@ bool checkreturn pb_encode_varint(pb_ostream_t *stream, pb_uint64_t value)
         i++;
     }
     buffer[i-1] &= 0x7F; /* Unset top bit on last byte */
-    
     return pb_write(stream, buffer, i);
 }
 
@@ -631,7 +635,6 @@ bool checkreturn pb_encode_tag_for_field(pb_ostream_t *stream, const pb_field_t
         default:
             PB_RETURN_ERROR(stream, "invalid field type");
     }
-    
     return pb_encode_tag(stream, wiretype, field->tag);
 }
 
@@ -709,7 +712,7 @@ static bool checkreturn pb_enc_varint(pb_ostream_t *stream, const pb_field_t *fi
         value = *(const pb_int64_t*)src;
     else
         PB_RETURN_ERROR(stream, "invalid data_size");
-    
+        
     return pb_encode_varint(stream, (pb_uint64_t)value);
 }
 
